### ISSUE TITLE
latest master branch build failure

### ISSUE THREAD

**Issue by Jinming-Hu** [Azure/azure-sdk-for-cpp#742]
```
[ 15%] Built target azure-core
[ 25%] Built target azure-storage-common
[ 30%] Built target azure-storage-files-shares
[ 38%] Built target azure-storage-blobs
[ 46%] Built target azure-storage-files-datalake
[ 50%] Built target azure-storage-sample
[ 52%] Built target gtest
[ 54%] Built target gmock
[ 56%] Built target gtest_main
[ 76%] Built target azure-storage-test
[ 86%] Built target azure-core-test
[ 88%] Built target azure_hang
[ 89%] Linking CXX executable azure_core_storage_issue_248
/usr/bin/ld: ../../libazure-core.a(curl.cpp.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/build.make:105: sdk/core/azure-core/test/e2e/azure_core_storage_issue_248] Error 1
make[1]: *** [CMakeFiles/Makefile2:597: sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

**Comment by RickWinter** [comment]
We are unable to reproduce this failure.  If you have a repro, please provide details on the OS, targeted arch, and any build flags/options you are passing.

**Comment by Jinming-Hu** [comment]
```
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=ON
make -j
```

**Comment by Jinming-Hu** [comment]
Ubuntu 20.04 LTS
gcc version 9.3.0
cmake version 3.18.0

**Comment by Jinming-Hu** [comment]
I can also reproduce it on CentOS

**Comment by Jinming-Hu** [comment]
seems it's caused by not linking to pthread.

can core sdk try to link to pthread like this

```
set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
set(THREADS_PREFER_PTHREAD_FLAG TRUE)
find_package(Threads REQUIRED)
target_link_libraries(YOUR_TARGET Threads::Threads)
```

**Comment by ahsonkhan** [comment]
What are the versions of dependencies being used? Did you use vcpkg to install them? If not, can you please share how you used `apt-get` to get the necessary dependencies.

**Comment by Jinming-Hu** [comment]
to reproduce on Ubuntu 20.04

```
apt-get update
apt-get install g++ cmake git libcurl4-openssl-dev libxml2-dev libssl-dev
git clone https://github.com/Azure/azure-sdk-for-cpp.git
cd azure-sdk-for-cpp/
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=ON
cmake --build .
```


error message:

```
Scanning dependencies of target azure_core_storage_issue_248
[ 89%] Building CXX object sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/azure_core_storage_list_containers_sample.cpp.o
[ 90%] Linking CXX executable azure_core_storage_issue_248
/usr/bin/ld: ../../libazure-core.a(curl.cpp.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/build.make:86: sdk/core/azure-core/test/e2e/azure_core_storage_issue_248] Error 1
make[1]: *** [CMakeFiles/Makefile2:564: sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

**Comment by lordgamez** [comment]
I have the same problem when -DBUILD_TESING option is ON on Ubuntu 20.04. Command:
`mkdir build && cd build && cmake -DBUILD_TESTING=ON .. && make -j12`

Message:
```
[ 23%] Linking CXX executable azure_core_storage_issue_248
/usr/bin/ld: ../../libazure-core.a(curl.cpp.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/build.make:86: sdk/core/azure-core/test/e2e/azure_core_storage_issue_248] Error 1
make[1]: *** [CMakeFiles/Makefile2:565: sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_248.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 24%] Linking CXX executable azure_core_with_curl
[ 25%] Linking CXX executable azure_core_with_curl_stream
/usr/bin/ld: ../../libazure-core.a(curl.cpp.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_with_curl.dir/build.make:86: sdk/core/azure-core/test/e2e/azure_core_with_curl] Error 1
make[1]: *** [CMakeFiles/Makefile2:646: sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_with_curl.dir/all] Error 2
/usr/bin/ld: ../../libazure-core.a(curl.cpp.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_with_curl_stream.dir/build.make:86: sdk/core/azure-core/test/e2e/azure_core_with_curl_stream] Error 1
make[1]: *** [CMakeFiles/Makefile2:619: sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_with_curl_stream.dir/all] Error 2
[ 26%] Linking CXX static library libazure-identity.a
[ 27%] Linking CXX executable azure_core_storage_issue_249
[ 27%] Built target azure-identity
/usr/bin/ld: ../../libazure-core.a(curl.cpp.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_249.dir/build.make:86: sdk/core/azure-core/test/e2e/azure_core_storage_issue_249] Error 1
make[1]: *** [CMakeFiles/Makefile2:592: sdk/core/azure-core/test/e2e/CMakeFiles/azure_core_storage_issue_249.dir/all] Error 2
[ 28%] Linking CXX static library ../../../lib/libgtestd.a
[ 28%] Built target gtest
make: *** [Makefile:141: all] Error 2
```

**Comment by Jinming-Hu** [comment]
@RickWinter This is impacting our customers, can anybody from Core sdk take a look and fix it?

**Comment by lordgamez** [comment]
The solution by @Jinming-Hu seems to be working, I opened a PR #980 with the proposed code change.

**Comment by vhvb1989** [comment]
fixed by #980
